e3/src/entity.rs

```diff
@@ -0,0 +1,395 @@
+    }
+}
+
+pub(crate) struct EntityHarness
+{
+    id: String,
+    joined_environments: Arc<Mutex<HashMap<String, Notifier>>>,
+    affected_environments: Arc<Mutex<HashMap<String, Waker>>>,
+    out_chan: Arc<Mutex<Broadcaster<Effect>>>,
+    drop_notifier: Arc<Mutex<Signal>>,
+    shutdown_listener: Arc<Mutex<SignalRx>>,
+    terminate_chan: Arc<Mutex<(watch::Sender<bool>, watch::Receiver<bool>)>>,
+    waker: Waker,
+    num_received_effects: Arc<AtomicUsize>,
+    entity: Arc<Mutex<Box<dyn Entity>>>
+}
+
+impl EntityHarness
+{
+    pub(crate) fn new(entity: Box<dyn Entity>, shutdown_rx: SignalRx) -> Self
+    {
+        Self {
+            id: Uuid::new_v4().to_string(),
+            joined_environments: share_mut!(HashMap::new()),
+            affected_environments: share_mut!(HashMap::new()),
+            out_chan: share_mut!(Broadcaster::new(BROADCAST_BUFFER_SIZE)),
+            drop_notifier: share_mut!(Signal::new()),
+            shutdown_listener: share_mut!(shutdown_rx),
+            terminate_chan: share_mut!(watch::channel(false)),
+            waker: Waker::new(),
+            num_received_effects: share!(AtomicUsize::new(0)),
+            entity: share_mut!(entity)
+        }
+    }
+
+    pub(crate) fn connect_input(
+        &mut self,
+        environment_name: &str,
+        environment_rx: BroadcastReceiver<Effect>,
+        environment_drop_rx: SignalRx
+    ) -> Waker
+    {
+        let mut joined_environments = unlock!(self.joined_environments);
+
+        if joined_environments.contains_key(environment_name) {
+            panic!(ENTITY_ALREADY_JOINED_ERROR);
+        }
+
+        joined_environments.insert(
+            environment_name.into(),
+            Notifier {
+                effect_rx: environment_rx,
+                drop_rx: environment_drop_rx
+            }
+        );
+
+        self.waker.clone()
+    }
+
+    pub(crate) fn disconnect_input(&mut self, environment_name: &str)
+    {
+        unlock!(self.joined_environments).remove(environment_name);
+    }
+
+    pub(crate) fn get_connectors(&mut self) -> (BroadcastReceiver<Effect>, SignalRx)
+    {
+        let effect_rx = unlock!(self.out_chan).add_rx();
+        let drop_rx = unlock!(self.drop_notifier).add_rx();
+
+        (effect_rx, drop_rx)
+    }
+
+    pub(crate) fn connect_output(&mut self, environment_name: &str, environment_waker: Waker)
+    {
+        let mut affected_environments = unlock!(self.affected_environments);
+
+        if affected_environments.contains_key(environment_name) {
+            panic!(ENTITY_ALREADY_AFFECTS_ERROR);
+        }
+
+        affected_environments.insert(environment_name.into(), environment_waker);
+    }
+
+    pub(crate) fn disconnect_output(&mut self, environment_name: &str)
+    {
+        unlock!(self.affected_environments).remove(environment_name);
+    }
+
+    pub(crate) fn terminate(&self)
+    {
+        unlock!(self.drop_notifier).emit();
+
+        #[cfg(debug_assertions)]
```

Comment:
I've seen the following `debug_assertions`. Maybe these should be replaced by error handling afterwards.

Reply:
using code that is only compiled in debug builds (`debug_assertions`) and error handling are orthogonal. While you always want to have good error handling in production code, those lines decorated with `debug_assertions` are ignored in a release build. That's why it's one of the todos to use some configurable logging crate (log) instead. Rest assured that this is definitely something that will be reworked.